api/rest/bundle_handler.go

```diff
@@ -0,0 +1,354 @@
+package rest
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/dcos/dcos-diagnostics/collectors"
+
+	"github.com/gorilla/mux"
+
+	"github.com/sirupsen/logrus"
+)
+
+// work dir contains only directories, each dir is created for single bundle (id is its name) and should contains:
+const (
+	stateFileName = "state.json" // file with information about diagnostics run
+	dataFileName  = "file.zip"   // data gathered by diagnostics
+
+	//TODO(janisz): Think about removing/replacing summary files with something more structured
+	summaryErrorsReportFileName = "summaryErrorsReport.txt" // error log in bundle
+	summaryReportFileName       = "summaryReport.txt"       // error log in bundle
+
+	filePerm = 0600
+	dirPerm  = 0700
+)
+
+type Bundle struct {
+	ID      string    `json:"id,omitempty"`
+	Size    int64     `json:"size,omitempty"` // length in bytes for regular files; 0 when Canceled or Deleted
+	Status  Status    `json:"status"`
+	Started time.Time `json:"started_at,omitempty"`
+	Stopped time.Time `json:"stopped_at,omitempty"`
+	Errors  []string  `json:"errors,omitempty"`
+}
+
+type ErrorResponse struct {
+	Code  int    `json:"code"`
+	Error string `json:"error"`
+}
+
+type Clock interface {
+	Now() time.Time
+}
+
+type realClock struct{}
+
+func (realClock) Now() time.Time { return time.Now() }
+
+func NewBundleHandler(workDir string, collectors []collectors.Collector) BundleHandler {
+	return BundleHandler{
+		clock:      realClock{},
+		workDir:    workDir,
+		collectors: collectors,
+	}
+}
+
+// BundleHandler is a struct that collects all functions
+// responsible for diagnostics bundle lifecycle
+type BundleHandler struct {
+	clock      Clock
+	workDir    string                 // location where bundles are generated and stored
+	collectors []collectors.Collector // information what should be in the bundle
+}
+
+func (h BundleHandler) Create(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["uuid"]
+
+	_, err := h.getBundleState(id)
+	if err == nil {
+		writeJSONError(w, http.StatusConflict, fmt.Errorf("bundle %s already exists", id))
+		return
+	}
+
+	bundleWorkDir := filepath.Join(h.workDir, id)
+	err = os.MkdirAll(bundleWorkDir, dirPerm)
+	if err != nil {
+		writeJSONError(w, http.StatusInsufficientStorage, fmt.Errorf("could not Create bundle %s workdir: %s", id, err))
```

Comment:
I used 507 to distinguishes workdir/permisions/storage configuration problems from other application related issues

> The method could not be performed on the resource because the server is unable to store the representation needed to successfully complete the request.

Reply:
But is that something we really need? Would the client behave differently than when it receives a simple 500? If not, just use 500. And we should log this fact in the diagnostics logs.

Reply:
Client could behave differently. This will reduce customer issues if we can display clear message when CLI gets 507 then just generic message when there is 500.